### PR TITLE
meta-iotqa:  add qa tests for mraa-gpio.

### PR DESCRIPTION
**Pull request by AshwinSoman**
Add tests for mraa-gpio python interface.

Signed-off-by: Ashwin Soman <ashwinx.soman@intel.com>

### CODE REVIEW

**Comment by AshwinSoman** [comment]
Please help to test this build.

**Comment by okartau** [comment]
ok to test

**Comment by okartau** [comment]
error from CI is not shown in main log, but it can be found deeper in aft log, specifically in
test_aft.minnowboardturbot.refkit-image-gateway.log.
There seems to be issue with comment that's not formatted as comment:

  File "iottest/oeqa/runtime/peripherals/mraa/mraa_commands.py", line 22
    copy all necessary files for test to the target device
          ^
SyntaxError: invalid syntax

**Comment by AshwinSoman** [comment]
I have resolved error and added few comments , please test

**Comment by okartau** [comment]
test_aft.570x.refkit-image-gateway.log:
  File "/root/workspace/workspace/intel-iot-refkit_pull-requests_6/iottest/oeqa/runtime/peripherals/mraa/mraa_commands.py", line 33
    '''
      ^
IndentationError: expected an indented block

**Comment by AshwinSoman** [comment]
Indentation Error resolved please test again

**Comment by AshwinSoman** [comment]
can I get , what is the output getting printed ?
or log file ?

**Comment by AshwinSoman** [comment]
I have resolved error please test .

**Comment by okartau** [comment]
there is no change in your PR content, you specifically want CI to test with same content again?
Or did you plan to push some change?

**Comment by AshwinSoman** [comment]
Hello

I have made following changes

4  meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_commands.py<https://github.com/intel/intel-iot-refkit/pull/254/commits/a1cdcb12739ca3cca99a6938bd63cdc4db31e65d#diff-91d582dfcae3c92d7865ed442e4b4cff>

@@ -19,8 +19,8 @@ class MraaGPIOTest(oeRuntimeTest):



     # target locations


     mraa_cmd_test_python_script_target = '/tmp/%s' % mraa_cmd_test_python_script


-    mraa_get_test_script_target = '/tmp/%s' % mraa-get_test_script


-    mraa_set_test_script_target = '/tmp/%s' % mraa-set_test_script


+    mraa_get_test_script_target = '/tmp/%s' % mraa_get_test_script


+    mraa_set_test_script_target = '/tmp/%s' % mraa_set_test_script


     mraa_list_test_script_target = '/tmp/%s' % mraa_list_test_script


     mraa_version_test_script_target = '/tmp/%s' % mraa_version_test_script




I want to test it

Regards
Ashwin

**Comment by okartau** [comment]
The change that you pointed:
commit a1cdcb12739ca3cca99a6938bd63cdc4db31e65d
was tested in most recent build 94 and failed, as indicated with red cross mark
next to commit-id in Github.
After that, there has been no new changes in this PR.
2 test failures is reported on job 94 front page.
I verified in job workspace, file contents was changed as you stated:
+    mraa_get_test_script_target = '/tmp/%s' % mraa_get_test_script
+    mraa_set_test_script_target = '/tmp/%s' % mraa_set_test_script

You want to retest same code?

**Comment by AshwinSoman** [comment]
Hello I have resolved that error it was bcz of a spelling mistake , plz test it again . thanks

File: meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_commands.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+    # target locations
+    mraa_cmd_test_python_script_target = '/tmp/%s' % mraa_cmd_test_python_script
+    mraa_get_test_script_target = '/tmp/%s' % mraa_get_test_script
+    mraa_set_test_script_target = '/tmp/%s' % mraa_set_test_script
+    mraa_list_test_script_target = '/tmp/%s' % mraa_list_test_script
+    mraa_version_test_script_target = '/tmp/%s' % mraa_version_test_script
+
+
+    def SetUp(self):
```
Comment (okartau) [inline meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_commands.py]: should this be called setUp(self) ?

seems it is not called prior to testong, causing lack of files in /tmp

**Comment by AshwinSoman** [comment]
I have removed copying of script can you please test and let me know the results

**Comment by okartau** [comment]
seems this approach does not work either.
But why you dont want to fix what seems like spelling mistake that I pointed out
in my previous comment: you have callled setup handler SetUp(),
while in other tests (which work) this handler is called setUp(),
also documentation that I found suggest this handler as setUp().

**Comment by AshwinSoman** [comment]
have changed SetUp to setUP please test .

**Comment by AshwinSoman** [comment]
resolved target copy error please test

**Comment by AshwinSoman** [comment]
have made paths from relative to absolute please check

**Comment by AshwinSoman** [comment]
please test changed path for call of shell scripts to absolute from relative

**Comment by AshwinSoman** [comment]
I have re-based and squashed my commits can you please review and merge it to master

**Comment by AshwinSoman** [comment]
please remove test build lable too .

**Comment by AshwinSoman** [comment]
I used shell script as I required output of various linux commands which I can easily save in a variable in a shell script . 
where as to save it in python takes more effort.

for example
in linux I have used  
pin_exported=$(diff /tmp/initialy_exported_gpio_pins.txt /tmp/exported_gpio_pins.txt | grep +gpio)
pin_exported=$(echo ${pin_exported:1})
to do it in python I have to 
p = subprocess.Popen(["diff", "/tmp/initialy_exported_gpio_pins.txt", "/tmp/exported_gpio_pins.txt",  "|", "grep", "+gpio"], stdout=subprocess.PIPE)
(output, err) = p.communicate()
p = subprocess.Popen([], stdout=subprocess.PIPE)


I can replace os.system with subprocess , no problem with that .

**Comment by pohly** [comment]
On Mon, 2017-07-31 at 09:30 +0000, Ashwin Soman wrote:
> for example
> in linux I have used
> pin_exported=$(diff /tmp/initialy_exported_gpio_pins.txt
> /tmp/exported_gpio_pins.txt | grep +gpio)
> pin_exported=$(echo ${pin_exported:1})
> to do it in python I have to
> p = subprocess.Popen(["diff", "/tmp/initialy_exported_gpio_pins.txt",
> "/tmp/exported_gpio_pins.txt", "|", "grep", "+gpio"],
> stdout=subprocess.PIPE)
> (output, err) = p.communicate()
> p = subprocess.Popen([], stdout=subprocess.PIPE)

You can use subprocess differently to make it shorter:

   pin_exported = subprocess.check_output("diff /tmp/initialy_exported_gpio_pins.txt /tmp/exported_gpio_pins.txt | grep +gpio",
                                          shell=True).decode('utf-8')

Using grep is one of the things I'm not fond of. You are not very
precise here, with the result that "+gpio" anywhere in the output will
satisfy the test. A stricter version of the test would look for a line
which contains exactly that word.

This can be fixed by extending the grep expression, but I argue that
doing such checks in Python would be clearer.

**Comment by ipuustin** [comment]
You shouldn't edit the commits that are already present the same PR. Just rebase the commits into one or two self-contained commits. Does this PR require there to be any special setup on the test device, such as connected GPIO ports? @Arthil and @samniem, can you tell if the tests get properly run and the test format is correct?

File: meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py [hunk 2]
```diff
@@ -18,30 +44,89 @@ def setUp(self):
...
+
+        status, output = self.target.run("mraa-gpio set " + self.pin + " 1")
         sleep(1)
-        gpio = check_gpio_output()
-        self.assertEqual(gpio, 0, msg="GPIO pin output is not 0, " +
-                                      "mraa output:\n" + output)
+        gpio = self.check_gpio_direction()
+        self.assertEqual(gpio, "out", msg="GPIO pin direction is not out, " +
+                                          "mraa output:\n" + gpio)
+        status, output = self.target.run("mraa-gpio get " + self.pin)
```
Comment (Arthil) [inline meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py]: I tried to manually test mraa-gpio get to see if it works but it doesn't. When setting the pin with mraa-gpio set I can't see any changes when running the mraa-gpio get command. For this reason, we are currently reading the value from Beaglebone's pin, which is connected to the pin we are changing. Do you have your own setup where you have tested the mraa-gpio get?

Comment (AshwinSoman) [inline meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py]: Even when I tested it manually by mra-gpio set first followed by mraa-gpio get on same board the value was not same but direction changes from out to in so I am not checking value , I am checking direction .
To check value for get I think we have to provide input from host board and the run get command and check for value .

**Comment by AshwinSoman** [comment]
I think It was not able to complete tests can any one help to find reason . As I am not able to  find any error . and even my test case didn't get executed

**Comment by Arthil** [comment]
Are you unable to see the logs generated in the CI? "test_aft.570x.refkit-image-gateway.log" gives the following error: 

File "/root/workspace/workspace/intel-iot-refkit_pull-requests_6/iottest/oeqa/runtime/peripherals/mraa/mraa_gpio.py", line 61
    print ("cmd_direction_at start")
                                   ^
TabError: inconsistent use of tabs and spaces in indentation

**Comment by okartau** [comment]
error is in one of daft log files, there are tabs used.
Traceback (most recent call last):
  File "iottest/runtest.py", line 256, in <module>
    ret = main()
  File "iottest/runtest.py", line 250, in main
    runner.loadTests()
  File "iottest/runtest.py", line 103, in loadTests
    super(RuntestTestContext, self).loadTests()
  File "/root/workspace/workspace/intel-iot-refkit_pull-requests_6/iottest/oeqa/oetest.py", line 298, in loadTests
    suites = [testloader.loadTestsFromName(name) for name in self.testslist]
  File "/root/workspace/workspace/intel-iot-refkit_pull-requests_6/iottest/oeqa/oetest.py", line 298, in <listcomp>
    suites = [testloader.loadTestsFromName(name) for name in self.testslist]
  File "/usr/lib/python3.4/unittest/loader.py", line 105, in loadTestsFromName
    module = __import__('.'.join(parts_copy))
  File "/root/workspace/workspace/intel-iot-refkit_pull-requests_6/iottest/oeqa/runtime/peripherals/mraa/mraa_gpio.py", line 61
    print ("cmd_direction_at start")
                                   ^
TabError: inconsistent use of tabs and spaces in indentation

**Comment by AshwinSoman** [comment]
I have restarted test by squashing my commits but please help me find error in https://intel-iot-refkit.01.org/job/intel-iot-refkit_pull-requests/221/

**Comment by okartau** [comment]
Have you considered verifying python syntax separately before pushing to GH?
Local test for "tabs and spaces" type of error is quick, does not even need modules imported,
example. I just copied mraa_gpio.py alone and can see syntax error in 2 easy ways:
======== method 1 ===========
$ pylint mraa_gpio.py 
No config file found, using default configuration
************* Module mraa_gpio
E: 61, 0: inconsistent use of tabs and spaces in indentation (syntax-error)

======== method 2 ===========
~ $ python3 mraa_gpio.py 
  File "mraa_gpio.py", line 61
    print ("cmd_direction_at start")
                                   ^
TabError: inconsistent use of tabs and spaces in indentation

That's much quicker than waiting 2+ hours for CI giving you same result

**Comment by AshwinSoman** [comment]
yes I am using pycharm to check it before pushing it , pycharm showed no error . I checked that particular line on pycharm again after getting error to check whether I missed it earlier but it was still showing  ok over there .
I have removed that line as print was just for sake of debugging 

And thanks for help in finding error as I was checking only test results and console output for error.

**Comment by okartau** [comment]
Traceback which is written only in daft log and not in main log is indeed a CI usability issue.
I have addressed that in another PR#275, plan is to get Tracebacks exposed in main console log

**Comment by AshwinSoman** [comment]
I have re-based and squashed my commits can you please review and merge it to master

**Comment by epyeoh** [comment]
@okartau hi olev, thanks for suggestion to use pylint to check indentation and other error. We will start using pylint, it was very useful!

**Comment by AshwinSoman** [comment]
please help to review and merge it to master

**Comment by mythi** [comment]
I'm confused about these changes and the justification so I need to ask:

- what's the purpose of the external pin checker script? The test does not seem to need that since the GPIO numbers used are fixed
- the "get" test simply checks that mraa sets the GPIO direction to "in" (and "out" after a set cmd) but ignores the value itself. how that justifies reading the GPIO value works?

**Comment by AshwinSoman** [comment]
External Pin checker script is there to find gpio node which gets created in DUT . and node number will be different for different boards so we need script to find the node which is created .

for reading values for get we have to set Beaglebone to out and send value and get it and check on DUT. 
I am working on that too . I will add another test case which will check value 
right now this test case checks weather the direction gets changed when we call get command or not.

**Comment by AshwinSoman** [comment]
please help to review and merge it to master

File: meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py [hunk 3]
```diff
@@ -18,30 +34,75 @@ def setUp(self):
...
+    def check_gpio_value(self):
+        cmd = "cat /sys/class/gpio/gpio20/value".split()
+        output = subprocess.check_output(cmd, stderr=subprocess.STDOUT)
+        return int(output)
+
+    def mraa_gpio_get_value(self):
+        status, output = self.target.run("mraa-gpio get " + self.pin)
+        output = str(output)
+        string1, string2 = output.split("=")
+        string2 = int(re.search(r'\d+', string2).group())
```
Comment (mythi) [inline meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py]: why is this needed? I believe you should also check ```status``` is OK before returning anything.

File: meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py [hunk 4]
```diff
@@ -18,30 +34,75 @@ def setUp(self):
...
+        it is shorted to gpio pin numeber 20 in BeagleBone
         '''
-        def check_gpio_output():
-            cmd = "cat /sys/class/gpio/gpio20/value".split()
-            output = subprocess.check_output(cmd, stderr=subprocess.STDOUT)
-            return int(output)
+        output = os.path.exists("/sys/class/gpio/gpio20")
+        output = str(output)
+        output = output.lower()
+        if "false" in output:
```
Comment (mythi) [inline meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py]: You can do ```if not os.path.exists("/sys/class/gpio/gpio20"):``` Also, the checking and exporting could be part of ```setUp```

File: meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py (hunk 3 above)
Comment (AshwinSoman) [inline meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py]: This part reads the output of mraa-gpio get <pin number> on DUT 
which will be like PIN 25 = 1 
I am extracting value from string and returning it .
I am writing value from Host and reading it on DUT using above command .
I am adding status check for self.target.run("mraa-gpio get " + self.pin)

File: meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py [hunk 5]
```diff
@@ -18,30 +34,75 @@ def setUp(self):
         else:
             raise unittest.SkipTest(output)
 
-    def test_gpio(self):
+    def check_gpio_value(self):
```
Comment (AshwinSoman) [inline meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py]: ok renaming it to host_gpio_write

**Comment by AshwinSoman** [comment]
please help to review and merge it to master

File: meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py (hunk 3 above)
Comment (mythi) [inline meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py]: I was referring to the ```re.search()``` part. There are naturally many was to do these things but something like

```python
_, string2 = output.split("=")
return int(string2)
```
would be suffient.

File: meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py [hunk 6]
```diff
@@ -18,30 +34,69 @@ def setUp(self):
...
+        subprocess.check_output("echo out > /sys/class/gpio/gpio20/direction", shell=True)
+        subprocess.check_output("echo 0 > /sys/class/gpio/gpio20/value", shell=True)
         sleep(1)
-        gpio = check_gpio_output()
-        self.assertEqual(gpio, 0, msg="GPIO pin output is not 0, " +
-                                      "mraa output:\n" + output)
+        output = self.mraa_gpio_get_value()
+        self.assertEqual(output, 0, msg="GPIO value is not 0 ")
+
+        subprocess.check_output("echo 1 > /sys/class/gpio/gpio20/value", shell=True)
```
Comment (mythi) [inline meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py]: I was asking to replace these with a ```host_gpio_write(value)``` function

File: meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py [hunk 7]
```diff
@@ -18,30 +34,69 @@ def setUp(self):
...
         self.assertEqual(gpio, 1, msg="GPIO pin output is not 1, " +
                                       "mraa output:\n" + output)
 
-        status, output = self.target.run("mraa-gpio set " + self.pin + " 0")
+    def test_gpio_cmd_list(self):
+
+        (status, output) = self.target.run("mraa-gpio list")
+        self.assertEqual(status, 0, msg="mraa-gpio list command failed: %s " % output)
+
+    def test_mraa_gpio_get_value(self):
```
Comment (mythi) [inline meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py]: one note about the ```test_``` function naming too. please make them consistent (e.g., ```test_mraa_gpio_cmd_<command>```

**Comment by AshwinSoman** [comment]
please help to review and merge it to master

File: meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py [hunk 8]
```diff
@@ -1,15 +1,30 @@
+'''
+this test suit tests functioning of four options mraa-gpio command
+mraa-gpio version
+mraa-gpio list
+mraa-gpio set <pin> <value>
+mraagpio get <pin>
```
Comment (mythi) [inline meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py]: typo

File: meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py [hunk 9]
```diff
@@ -1,15 +1,30 @@
...
+'''
+
 from oeqa.oetest import oeRuntimeTest
 import unittest
 import subprocess
+import os
+import re
 from time import sleep
 
+
```
Comment (mythi) [inline meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py]: cleanup these please

File: meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py [hunk 10]
```diff
@@ -18,30 +33,72 @@ def setUp(self):
...
+    def host_gpio_write(self, value):
+        cmd = 'echo %s > /sys/class/gpio/gpio20/value' % value
+        subprocess.check_output(cmd, shell=True)
+
+    def dut_mraa_gpio_get_value(self):
+        status, output = self.target.run("mraa-gpio get " + self.pin)
+        output = str(output)
+        self.assertEqual(status, 0, msg="mraa-gpio get command failed: %s " % output)
+        string1, string2 = output.split("=")
+        string2 = int(re.search(r'\d+', string2).group())
```
Comment (mythi) [inline meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py]: my earlier feedback is not addressed here

File: meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py [hunk 11]
```diff
@@ -18,30 +33,72 @@ def setUp(self):
         else:
             raise unittest.SkipTest(output)
 
-    def test_gpio(self):
+        if not (str(os.path.exists("/sys/class/gpio/gpio20"))):
```
Comment (mythi) [inline meta-iotqa/lib/oeqa/runtime/peripherals/mraa/mraa_gpio.py]: just ```if not os.path...``` (drop parentheses)

**Comment by AshwinSoman** [comment]
@mythi  can you restart test for this commit as the error coming is irrelevant to my changes .

**Comment by mythi** [comment]
test this please
